src/System.Text.Primitives/System/Text/Formatting/Internal/IntegerFormatter.Utf16.cs

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    internal static class InvariantUtf16IntegerFormatter
+    {
+        private const char Minus = '-';
+        private const char Period = '.';
+        private const char Seperator = ',';
+
+        private const int GroupSize = 3;
+
+        public static bool TryFormatDecimalInt64(long value, byte precision, Span<byte> buffer, out int bytesWritten)
+        {
+            int digitCount = FormattingHelpers.CountDigits(value);
+            int charsNeeded = digitCount + ((value < 0) ? 1 : 0);
+            Span<char> span = buffer.NonPortableCast<byte, char>();
+
+            if (span.Length < charsNeeded)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            ref char utf16Bytes = ref span.DangerousGetPinnableReference();
+            int idx = 0;
+
+            if (value < 0)
+            {
+                Unsafe.Add(ref utf16Bytes, idx++) = Minus;
+
+                // Abs(long.MinValue) == long.MaxValue + 1, so we need to re-route to unsigned to handle value
+                if (value == long.MinValue)
+                {
+                    if (!TryFormatDecimalUInt64((ulong)long.MaxValue + 1, precision, buffer.Slice(2), out bytesWritten))
+                        return false;
+
+                    bytesWritten += sizeof(char); // Add the minus sign
+                    return true;
+                }
+
+                value = -value;
+            }
+
+            if (precision != TextFormat.NoPrecision)
+            {
+                int leadingZeros = (int)precision - digitCount;
+                while (leadingZeros-- > 0)
+                    Unsafe.Add(ref utf16Bytes, idx++) = '0';
+            }
+
+            idx += FormattingHelpers.WriteDigits(value, digitCount, ref utf16Bytes, idx);
+
+            bytesWritten = idx * sizeof(char);
+            return true;
+        }
+
+        public static bool TryFormatDecimalUInt64(ulong value, byte precision, Span<byte> buffer, out int bytesWritten)
+        {
+            if (value <= long.MaxValue)
+                return TryFormatDecimalInt64((long)value, precision, buffer, out bytesWritten);
+
+            // Remove a single digit from the number. This will get it below long.MaxValue
+            // Then we call the faster long version and follow-up with writing the last
+            // digit. This ends up being faster by a factor of 2 than to just do the entire
+            // operation using the unsigned versions.
+            value = FormattingHelpers.DivMod(value, 10, out ulong m);
```

Comment:
At that point, the implementation becomes entirely unsigned with the signed version printing a minus when necessary and calling into unsigned with absolute value.